### PR TITLE
Adding Tpr info and time in attendance CSV

### PR DESCRIPTION
**Pull request by SakshiSharan**
Adding info of tpr who took attendance and time at which attendance is taken in attendance CSV

### CODE REVIEW

File: controller/attendance.go [hunk 1]
```diff
@@ -114,13 +114,59 @@ func GetAttendanceByExaminationIdExport(ctx *gin.Context, client *mongo.Client,
...
+		for _, markedBy := range student.MarkedData {
+			tpr[markedBy.User] = true
+		}
+	}
+
+	for k, v := range tpr {
+		if v == true {
+			tprIDs = append(tprIDs, k)
+		}
+	}
```
Comment (Shubhrajyoti-Dey-FrosTiK) [inline controller/attendance.go:130]: The whole could have been done in one loop

```.go
var tpr map[primitive.ObjectID]bool = make(map[primitive.ObjectID]bool)
var tprIDs []primitive.ObjectID

	for _, student := range attendance {
		for _, markedBy := range student.MarkedData {
                        if tpr[markedBy.User]  == false {
                              tprIDs = append(tprIDs, k)
			tpr[markedBy.User.User] = true
                       }
                      tpr[markedBy.User] = true
		}
	}

```

File: controller/attendance.go [hunk 2]
```diff
@@ -114,13 +114,59 @@ func GetAttendanceByExaminationIdExport(ctx *gin.Context, client *mongo.Client,
...
+		[]bson.M{
+			{
+				"$match": bson.M{
+					"_id": bson.M{
+						"$in": tprIDs,
+					},
+				},
+			},
+		},
+	)
```
Comment (Shubhrajyoti-Dey-FrosTiK) [inline controller/attendance.go:144]: This should not be `Aggregate`

`Find` method should be used here.

File: controller/attendance.go [hunk 3]
```diff
@@ -114,13 +114,59 @@ func GetAttendanceByExaminationIdExport(ctx *gin.Context, client *mongo.Client,
...
+		return nil, err
+	}
+	cursor.All(ctx, &tprs)
+
 	for _, student := range attendance {
 		studentDetails := student.UserDetails[0]
+		var takenBy string = ""
+		for _, markedBy := range student.MarkedData {
+			for _, tpr := range tprs {
+				if tpr.Id == markedBy.User {
```
Comment (Shubhrajyoti-Dey-FrosTiK) [inline controller/attendance.go:156]: A map should be used here.

Don't loop tpr every time

Should be like this.

```.go

// Store objectID -> Full Name
var tprMap map[primitive.ObjectID]string = make(map[primitive.ObjectID]string)

for _, tpr := range tprs {
     tprMap[tpr.Id] = utils.GetName(tpr.FirstName, tpr.MiddleName, tpr.LastName)
}


// Use this map afterwards like
var takenBy string = tprMap[markedBy.User] + " (" + markedBy.Time + ")"
```

**Review by Shubhrajyoti-Dey-FrosTiK — CHANGES_REQUESTED** [review]

File: controller/attendance.go (hunk 1 above)
Comment (SakshiSharan) [reply controller/attendance.go]: Did the changes as above and tested.

File: controller/attendance.go (hunk 2 above)
Comment (SakshiSharan) [reply controller/attendance.go:136]: Used Find and tested the code.

File: controller/attendance.go (hunk 3 above)
Comment (SakshiSharan) [reply controller/attendance.go]: Changed code as above and tested.

File: controller/attendance.go [hunk 4]
```diff
@@ -114,13 +114,53 @@ func GetAttendanceByExaminationIdExport(ctx *gin.Context, client *mongo.Client,
...
+
+	var tprName map[primitive.ObjectID]string = make(map[primitive.ObjectID]string)
+
+	for _, tpr := range tprs {
+		tprName[tpr.Id] = utils.GetName(tpr.FirstName, tpr.MiddleName, tpr.LastName)
+	}
+
 	for _, student := range attendance {
 		studentDetails := student.UserDetails[0]
+		var takenBy string = ""
```
Comment (Shubhrajyoti-Dey-FrosTiK) [inline controller/attendance.go:151]: Update it like this

```.go
var markedByList []string = make([]string,0)

for _, student := range attendance {
    markedByList = append(markedByList, tprName[markedBy.User] + " (" + markedBy.Time + ")")
}

// And in the bottom write
TakenBy: strings.Join(markedByList, " , ")

```

**Review by Shubhrajyoti-Dey-FrosTiK — CHANGES_REQUESTED** [review]

File: controller/attendance.go (hunk 4 above)
Comment (SakshiSharan) [reply controller/attendance.go]: Resolved as above and tested.

**Review by Shubhrajyoti-Dey-FrosTiK — APPROVED** [review]

**Pull request merged**